packages/web3-eth-contract/src/contract.ts

```diff
@@ -347,25 +347,40 @@ export class Contract<Abi extends ContractAbi>
 		contextOrReturnFormat?: Web3ContractContext | DataFormat,
 		returnFormat?: DataFormat,
 	) {
+		let contractContext;
+		if (isWeb3ContractContext(addressOrOptionsOrContext)) {
+			contractContext = addressOrOptionsOrContext;
+		} else if (isWeb3ContractContext(optionsOrContextOrReturnFormat)) {
+			contractContext = optionsOrContextOrReturnFormat;
+		} else {
+			contractContext = contextOrReturnFormat;
+		}
+
+		let provider;
+		if (
+			typeof addressOrOptionsOrContext === 'object' &&
+			'provider' in addressOrOptionsOrContext
+		) {
+			provider = addressOrOptionsOrContext.provider;
+		} else if (
+			provider === undefined &&
+			typeof optionsOrContextOrReturnFormat === 'object' &&
+			'provider' in optionsOrContextOrReturnFormat
+		) {
+			provider = optionsOrContextOrReturnFormat.provider;
+		} else if (
+			provider === undefined &&
+			typeof contextOrReturnFormat === 'object' &&
+			'provider' in contextOrReturnFormat
+		) {
+			provider = contextOrReturnFormat.provider;
+		} else if (provider === undefined && Contract.givenProvider !== undefined) {
+			provider = Contract.givenProvider;
+		}
```

Comment:
Actually, provider === undefined is always `true` inside those `else if` statements. Since the variable is never being initialized before those conditions. 
Another point is that the last condition for `else if` is useless. And it could be just `else`: 

```suggestion
		} else if (
			typeof optionsOrContextOrReturnFormat === 'object' &&
			'provider' in optionsOrContextOrReturnFormat
		) {
			provider = optionsOrContextOrReturnFormat.provider;
		} else if (
			typeof contextOrReturnFormat === 'object' &&
			'provider' in contextOrReturnFormat
		) {
			provider = contextOrReturnFormat.provider;
		} else {
			provider = Contract.givenProvider;
		}
```

Except if we need to do something if `Contract.givenProvider` also `undefined`:
```suggestion
		} else if (
			typeof optionsOrContextOrReturnFormat === 'object' &&
			'provider' in optionsOrContextOrReturnFormat
		) {
			provider = optionsOrContextOrReturnFormat.provider;
		} else if (
			typeof contextOrReturnFormat === 'object' &&
			'provider' in contextOrReturnFormat
		) {
			provider = contextOrReturnFormat.provider;
		} else if (Contract.givenProvider !== undefined) {
			provider = Contract.givenProvider;
		} else {
		    //... in case there is something was want to do in this case...
		}
```

Reply:
Thank you, I was thinking about this as if they were just if statements

I think 

```typescript
else {
    provider = Contract.givenProvider;
}
```

should be acceptable as I believe that was what the code before the refactor was intending

Reply:
Addressed by [this](https://github.com/web3/web3.js/pull/5669/commits/d1b0b83adb6845ff0b72650048154a9ffb24fd0a) commit